auth/design.md

```diff
@@ -0,0 +1,69 @@
+﻿
+# Design document of the authentication mechanism in etcd v3 protocol
+
+## Why not using the existing v2 auth mechanism?
+The v3 protocol uses gRPC as its transport instead of RESTful interface like v2. Therefore we have a chance of improvements (e.g. connection based auth instead of request based auth). In addition, there is an unclear point in the semantics of the v2 auth (described later). In v3, it should be cleaned and a well defined guidance should be provided for users and developers.
+
+### Functionality requirements
+1. per connection authentication, not per request
+   1. User ID + password based authentication implemented on gRPC API
+   1. authentication must be refreshed after auth information update
+1. its functionality should be as simple and useful as v2
+   1. v3 provides flat key space, unlike the directory structure of v2. Permission check will be provided as simple prefix matching.
+1. it should avoid the consistency problem of v2, described later
+
+### Main required changes
+1. a client needs to create a dedicated connection just for authentication before sending actual requests
+1. add permission information (user ID and authorized revision) to the Raft commands (`etcdserverpb.InternalRaftRequest`)
+1. every request requires permission check in the state machine layer, not in the API layer
+
+### Consistency of the metadata for auth
+The metadata for auth should also be stored and managed in the storage controlled by etcd's Raft protocol like other data stored in etcd. It is required for not sacrificing availability and consistency of the entire etcd cluster. If reading or writing the metadata (e.g. permission information) needs an agreement of every node (more than quorum), single node failure can stop the entire cluster. Because permission checking during ordinal reading/writing requests cannot be completed if the cluster lacks a single node, even if it has living quorum nodes. It degrades availability of the cluster.

```

Comment:
s/Because... availability of the cluster/
Requiring all nodes to agree at once means that checking ordinary read/write requests cannot be completed if any cluster member is down, even if the cluster has an available quorum. This unanimous scheme ultimately degrades cluster availability; quorum based consensus from raft should suffice since agreement follows from consistent ordering.